File: java/0997.0_Find_the_Town_Judge.java
/*
执行用时：2 ms, 在所有 Java 提交中击败了99.21% 的用户
内存消耗：45.6 MB, 在所有 Java 提交中击败了93.00% 的用户
通过测试用例：92 / 92
*/

class Solution {
    public int findJudge(int n, int[][] trust) {
        int[] in = new int[n + 1], out = new int[n + 1];
        for (int[] t: trust) {
            int a = t[0], b = t[1];
            in[b]++;
            out[a]++;
        }
        for (int i = 1; i < n + 1; i++) {
            if (in[i] == n - 1 && out[i] == 0) {
                return i;
            }
        }
        return -1;
    }
}


/*
执行用时：2 ms, 在所有 Java 提交中击败了99.21% 的用户
内存消耗：46 MB, 在所有 Java 提交中击败了45.77% 的用户
通过测试用例：92 / 92
*/
class Solution {
    public int findJudge(int n, int[][] trust) {
        int[] in = new int[n + 1], out = new int[n + 1];
        for (int i = 0; i < trust.length; i++) {
            int a = trust[i][0], b = trust[i][1];
            in[b]++;
            out[a]++;
        }
        for (int i = 1; i < n + 1; i++) {
            if (in[i] == n - 1 && out[i] == 0) {
                return i;
            }
        }
        return -1;
    }
}


/*
Runtime: 4 ms, faster than 39.54% of Java online submissions for Find the Town Judge.
Memory Usage: 81.1 MB, less than 21.59% of Java online submissions for Find the Town Judge.
*/
class Solution {
    public int findJudge(int n, int[][] trust) {
        int[] degree = new int[n + 1];
        for (int[] t: trust) {
            degree[t[0]]--;
            degree[t[1]]++;
        }
        for (int i = 1; i <= n; ++i) {
            if (degree[i] == n - 1) {
                return i;
            }
        }
        return -1;
    }
}

File: problems/0997.0_Find_the_Town_Judge.py
'''
T: O(T + 3 * N)
S: O(2 * N)

Runtime: 1250 ms, faster than 5.72% of Python3 online submissions for Find the Town Judge.
Memory Usage: 18.9 MB, less than 65.18% of Python3 online submissions for Find the Town Judge.
'''
class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        candidates = set(range(1, n + 1))
        cnt = [0] * (n + 1)
        for a, b in trust:
            candidates.discard(a)
            cnt[b] += 1
        candidates = list(filter(lambda cand: cnt[cand] == n - 1, candidates))
        
        if len(candidates) == 1:
            return candidates[0]
        else:
            return -1


'''
directed graph
T: O(T + 3 * N)
S: O(2 * N)

Runtime: 948 ms, faster than 14.22% of Python3 online submissions for Find the Town Judge.
Memory Usage: 19 MB, less than 65.18% of Python3 online submissions for Find the Town Judge.
'''
class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        out_degree, in_degree = [0] * (n + 1), [0] * (n + 1)
        # a -> b
        for a, b in trust:
            out_degree[a] += 1
            in_degree[b] += 1
        for i in range(1, n + 1):
            if out_degree[i] == 0 and in_degree[i] == n - 1:
                return i
        return -1

'''
directed graph
T: O(T + 2 * N)
S: O(N)

Runtime: 1449 ms, faster than 5.04% of Python3 online submissions for Find the Town Judge.
Memory Usage: 19 MB, less than 65.18% of Python3 online submissions for Find the Town Judge.
'''
class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        # out_degree, in_degree = [0] * (n + 1), [0] * (n + 1)
        # in_degree - out_degree == n - 1
        degree = [0] * (n + 1)
        # a -> b
        for a, b in trust:
            degree[a] -= 1
            degree[b] += 1
        for i in range(1, n + 1):
            if degree[i] == n - 1:
                return i
        return -1

